### PR TITLE
Add seeding to dataset creation, remove label maps

### LINKED ISSUE #17: Randomness in build_tfrecords.py + label maps
Hi @avital , first of all thanks for releasing the code, I find it very elegant, and tmuxp sessions for every experiment are really cool!

There is information in the README that in order to have exactly the same labeled/unlabeled divisions as in the paper, one has to use label maps from the repo. But the issue is, in order to run the experiments one has to first build tfrecords, which uses random shuffle. This means we HAVE TO create new label maps for these newly created tfrecords.
Am I missing something, or there is really an issue here? How can we generate exactly the same tfrecords as in your experiments?

### LINKED ISSUE #18: README says to build_tfrecords.py for imagenet_32 before downloading imagenet_32
The section "Prepare datasets" includes the code snippet
> `python3 build_tfrecords.py --dataset_name=imagenet_32`

which appears before the section describing how to download imagenet_32.

### PR DESCRIPTION
**Pull request by craffel**
Resolves #17 and #18. It was impossible to use the existing label maps because the datasets needed to be downloaded, which necessitated creating new label maps. Instead, we should just make the dataset and label map creation processes deterministic by manually setting seeds. This involved changing the README to include label map creation, so I also fixed #18 which specified ImageNet 32x32 creation in the wrong order.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,22 +14,17 @@ The latest version of this repository can be found
 # Prepare datasets
 
 For SVHN and CIFAR-10, we provide scripts to automatically download and preprocess the data.
+We also provide a script to create "label maps", which specify which entries of the dataset should be treated as labeled and unlabeled.
```
Comment (avital) [inline README.md]:
```suggestion
We also provide a script to create "label maps", which specify which entries of the dataset should be treated as labeled and unlabeled. Both of these scripts use an explicitly chosen random seed, so the same dataset order and label maps will be created each time. The random seeds can be overridden, for example to test robustness to different labeled splits.
```

**Comment by avital** [comment]
Code LGTM. Did you test that both build_tfrecords.py and build_label_maps.py now generate the same results when the seed is fixed?

**Comment by avital** [comment]
Do you happen to know if Python and numpy guarantee reliability when upgrading to new versions?

**Review by avital — CHANGES_REQUESTED** [review]
I suggested a small change to the README, and I asked a question about testing reproducibility. Other than that LGTM.

**Comment by avital** [comment]
Looks like https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.random.RandomState.html has explicit guarantees which seem stronger than calling into random directly. I don't feel too strongly about this but could be worth switching to that...

**Comment by avital** [comment]
Specifically, it's possible that a library that we import calls random unknowingly to us, or change when/whether they call random in a version upgrade. Using an explicit RandomState would relieve us of that (tiny) risk of non-determinism when upgrading dependencies.

**Comment by craffel** [comment]
> Did you test that both build_tfrecords.py and build_label_maps.py now generate the same results when the seed is fixed?

Yes.

> Do you happen to know if Python and numpy guarantee reliability when upgrading to new versions?

Not sure.

> Specifically, it's possible that a library that we import calls random unknowingly to us, or change when/whether they call random in a version upgrade. Using an explicit RandomState would relieve us of that (tiny) risk of non-determinism when upgrading dependencies.

Ok, will make this change.

**Comment by craffel** [comment]
Ok, pushed a commit to switch to `np.random.RandomState`. The semantics in `build_label_maps.py` changed because of switching from `random.sample` to `np.random.choice`, PTAthoroughL.

File: build_label_map.py [hunk 2]
```diff
@@ -140,17 +135,21 @@ def build_single_label_map(
...
+            result_dict["values"] += [unique_ids[n] for n in random_ids]
     elif dataset_name in {"cifar10", "svhn", "cifar_unnormalized"}:
         path = os.path.join(fkeys_path, dataset_name, "label_to_fkeys_train")
         with gfile.GFile(path, "r") as f:
             label_to_fkeys = json.load(f)
         for label in label_to_fkeys.keys():
-            result_dict["values"] += random.sample(
-                label_to_fkeys[label], n_labeled_per_class
+            random_ids = rng.choice(
+                len(label_to_fkeys[label]), n_labeled_per_class, replace=False
```
Comment (avital) [inline build_label_map.py:148]: Wow, good catch to use `replace=False`.

**Review by avital — APPROVED** [review]

**Pull request merged**